game/src/main/scala/io/aigar/game/Cell.scala

```diff
@@ -59,6 +59,14 @@ class Cell(val id: Int, startPosition: Vector2 = new Vector2(0f, 0f)) {
 
     velocity += acceleration * deltaSeconds
     position += velocity * deltaSeconds
+
+    if (position.x <= 0 || position.x >= grid.width){
+      velocity = new Vector2(0f, velocity.y)
+    }
+    if (position.y <= 0 || position.y >= grid.height){
+      velocity = new Vector2(velocity.x, 0f)
+    }
```

Comment:
I think that's as clean as it gets. I'm okay with the way you did it! However, I would put this logic along with the `clamp` inside a method such as `keepInGrid`.